pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java

```diff
@@ -897,27 +898,54 @@ public CompletableFuture<Void> close() {
      */
     @Override
     public synchronized CompletableFuture<Void> disconnect() {
-        CompletableFuture<Void> disconnectFuture = new CompletableFuture<>();
+        if (fenceFuture != null && !fenceFuture.isDone()){
+            return fenceFuture;
+        }
+        if (fenceFuture != null && fenceFuture.isDone() && !fenceFuture.isCompletedExceptionally()){
+            return fenceFuture;
+        }
+        fenceFuture = new CompletableFuture<>();
 
         // block any further consumers on this subscription
         IS_FENCED_UPDATER.set(this, TRUE);
 
         (dispatcher != null ? dispatcher.close() : CompletableFuture.completedFuture(null))
                 .thenCompose(v -> close()).thenRun(() -> {
                     log.info("[{}][{}] Successfully disconnected and closed subscription", topicName, subName);
-                    disconnectFuture.complete(null);
+                    fenceFuture.complete(null);
                 }).exceptionally(exception -> {
                     IS_FENCED_UPDATER.set(this, FALSE);
                     if (dispatcher != null) {
                         dispatcher.reset();
                     }
                     log.error("[{}][{}] Error disconnecting consumers from subscription", topicName, subName,
                             exception);
-                    disconnectFuture.completeExceptionally(exception);
+                    fenceFuture.completeExceptionally(exception);
                     return null;
                 });
 
-        return disconnectFuture;
+        return fenceFuture;
+    }
+
+    /**
+     * Resume subscription after topic deletion or close failure.
+     */
+    public synchronized CompletableFuture<Void> resumeAfterFence() {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        fenceFuture.whenComplete((ignore, ignore2) -> {
+            try {
+                if (IS_FENCED_UPDATER.compareAndSet(this, TRUE, FALSE)) {
+                    if (dispatcher != null) {
+                        dispatcher.reset();
+                    }
+                }
+                result.complete(null);
+            } catch (Exception ex){
+                result.completeExceptionally(ex);
+                log.error("[{}] Resume subscription[{}] failure: {}", topicName, subName, ex.getMessage(), ex);
```

Comment:
please move this before completeExceptionally, otherwise logs may be hard to understand

Reply:
Already fixed